### PR TITLE
NrnState block transformations in AST to help code generation

### PR DESCRIPTION
**Pull request by pramodk**
- added Void node type to represent void return type
  - llvm helper started handling visit_nrn_state_block
    - NrnStateBlock is being converted into CodegenFunction
    - for loop body with solution blocks created
    - voltage and node index initialization code added

Todo

- [x] Handle ION updates statements
- [x] Handle breakpoint block in case of empty currents
- [x] Replace all range variable types with indexed variable

We have MOD file as:

```
 PROCEDURE rates(v(mV)) {
     LOCAL alpha, beta, sum, q10
     UNITSOFF
     q10 = 3^((celsius-6.3)/10)
     alpha = .1*vtrap(-(v+40), 10)
     beta = 4*exp(-(v+65)/18)
     sum = alpha+beta
     mtau = 1/(q10*sum)
     minf = alpha/sum
     alpha = .07*exp(-(v+65)/20)
     beta = 1/(exp(-(v+35)/10)+1)
     sum = alpha+beta
     htau = 1/(q10*sum)
     hinf = alpha/sum
     alpha = .01*vtrap(-(v+55), 10)
     beta = .125*exp(-(v+65)/80)
     sum = alpha+beta
     ntau = 1/(q10*sum)
     ninf = alpha/sum
 }

DERIVATIVE states {
         rates(v)
         m' =  (minf-m)/mtau
         h' = (hinf-h)/htau
         n' = (ninf-n)/ntau
 }
```

After various passes of NMODL, before code generation, MOD file AST is:

```
 PROCEDURE rates(v(mV)) {
     LOCAL alpha, beta, sum, q10
     UNITSOFF
     q10 = 3^((celsius-6.3)/10)
     alpha = .1*vtrap(-(v+40), 10)
     beta = 4*exp(-(v+65)/18)
     sum = alpha+beta
     mtau = 1/(q10*sum)
     minf = alpha/sum
     alpha = .07*exp(-(v+65)/20)
     beta = 1/(exp(-(v+35)/10)+1)
     sum = alpha+beta
     htau = 1/(q10*sum)
     hinf = alpha/sum
     alpha = .01*vtrap(-(v+55), 10)
     beta = .125*exp(-(v+65)/80)
     sum = alpha+beta
     ntau = 1/(q10*sum)
     ninf = alpha/sum
 }

 NRN_STATE SOLVE states METHOD cnexp{
     rates(v)
     m = m+(1.0-exp(dt*((((-1.0)))/mtau)))*(-(((minf))/mtau)/((((-1.0)))/mtau)-m)
     h = h+(1.0-exp(dt*((((-1.0)))/htau)))*(-(((hinf))/htau)/((((-1.0)))/htau)-h)
     n = n+(1.0-exp(dt*((((-1.0)))/ntau)))*(-(((ninf))/ntau)/((((-1.0)))/ntau)-n)
 }
```
Generated C++ code is:

```cpp
     void nrn_state_hh(NrnThread* nt, Memb_list* ml, int type) {
         int nodecount = ml->nodecount;
         int pnodecount = ml->_nodecount_padded;
         const int* __restrict__ node_index = ml->nodeindices;
         double* __restrict__ data = ml->data;
         const double* __restrict__ voltage = nt->_actual_v;
         Datum* __restrict__ indexes = ml->pdata;
         ThreadDatum* __restrict__ thread = ml->_thread;
         hh_Instance* __restrict__ inst = (hh_Instance*) ml->instance;

         int start = 0;
         int end = nodecount;
         #pragma ivdep
         for (int id = start; id < end; id++) {
             int node_id = node_index[id];
             double v = voltage[node_id];
             inst->ena[id] = inst->ion_ena[indexes[0*pnodecount+id]];
             inst->ek[id] = inst->ion_ek[indexes[3*pnodecount+id]];
             rates_hh(id, pnodecount, inst, data, indexes, thread, nt, v, v);
             inst->m[id] = inst->m[id] + (1.0 - exp(nt->_dt * (((( -1.0))) / inst->mtau[id]))) * ( -(((inst->minf[id])) / inst->mtau[id]) / (((( -1.0))) / inst->mtau[id]) - inst-  >m[id]);
             inst->h[id] = inst->h[id] + (1.0 - exp(nt->_dt * (((( -1.0))) / inst->htau[id]))) * ( -(((inst->hinf[id])) / inst->htau[id]) / (((( -1.0))) / inst->htau[id]) - inst-  >h[id]);
             inst->n[id] = inst->n[id] + (1.0 - exp(nt->_dt * (((( -1.0))) / inst->ntau[id]))) * ( -(((inst->ninf[id])) / inst->ntau[id]) / (((( -1.0))) / inst->ntau[id]) - inst-  >n[id]);
         }
     }
```

With this PR, transformed NMODL AST is:

```
 INTEGER rates(DOUBLE arg_v){
     INTEGER ret_rates                                     // note : this should be initialised to zero by default
     DOUBLE alpha, beta, sum, q10
     UNITSOFF
     q10 = 3^((celsius-6.3)/10)
     alpha = .1*vtrap(-(arg_v+40), 10)
     beta = 4*exp(-(arg_v+65)/18)
     sum = alpha+beta
     mtau[id] = 1/(q10*sum)
     minf[id] = alpha/sum
     alpha = .07*exp(-(arg_v+65)/20)
     beta = 1/(exp(-(arg_v+35)/10)+1)
     sum = alpha+beta
     htau[id] = 1/(q10*sum)
     hinf[id] = alpha/sum
     alpha = .01*vtrap(-(arg_v+55), 10)
     beta = .125*exp(-(arg_v+65)/80)
     sum = alpha+beta
     ntau[id] = 1/(q10*sum)
     ninf[id] = alpha/sum
     return ret_rates
 }

VOID nrn_state_hh(){
    INTEGER id
    for(id = 0; id<node_count; id = id+1) {
        INTEGER node_id, ena_id, ek_id
        DOUBLE v
        node_id = node_index[id]
        ena_id = ion_ena_index[id]
        ek_id = ion_ek_index[id]
        v = voltage[node_id]
        ena[id] = ion_ena[ena_id]
        ek[id] = ion_ek[ek_id]
        rates(v)
        m[id] = m[id]+(1.0-exp(dt*((((-1.0)))/mtau[id])))*(-(((minf[id]))/mtau[id])/((((-1.0)))/mtau[id])-m[id])
        h[id] = h[id]+(1.0-exp(dt*((((-1.0)))/htau[id])))*(-(((hinf[id]))/htau[id])/((((-1.0)))/htau[id])-h[id])
        n[id] = n[id]+(1.0-exp(dt*((((-1.0)))/ntau[id])))*(-(((ninf[id]))/ntau[id])/((((-1.0)))/ntau[id])-n[id])
    }
}
```

See #474

### CODE REVIEW

**Comment by pramodk** [comment]
@georgemitenkov : See `VOID nrn_state_hh()` and  `INTEGER rates(DOUBLE arg_v)` methods above. I think these now functions are in necessary form for you to take care for IR generation. Take a look and let me know if you think any additional changes will be helpful for you.

File: src/language/codegen.yaml [hunk 1]
```diff
@@ -216,4 +219,36 @@
                             type: CodegenVar
                             vector: true
                             separator: ", "
-                            add: true
\ No newline at end of file
+                            add: true
+                  - CodegenAtomicStatement:
+                      brief: "Represent atomic operation"
+                      description: |
+                        During code generation certain operations like ion updaates, vec_rhs or
```
Comment (georgemitenkov) [inline src/language/codegen.yaml:226]: nit: typo in updates

**Comment by georgemitenkov** [comment]
This LGTM, but I will leave that for @iomaganaris to approve :)

This looks great! Adding type information and transforming NMODL to the state kernel on the AST level helps the backends a lot! Then the actual codegen does not have to involve duplicate AST passes 👍

**Comment by georgemitenkov** [comment]
I will add a few changes when this PR lands to IR generation (basically `VOID` type support plus extracting type from variables rather than using default values). I think at the moment all necessary information is created.

**Review by iomaganaris — COMMENTED** [review]
LGTM in general. I have some left some small suggestions.
I also got a few more questions but we can discuss offline.

File: src/codegen/codegen_info.cpp [hunk 2]
```diff
@@ -131,5 +134,135 @@ bool CodegenInfo::is_voltage_used_by_watch_statements() const {
...
+}
+
+
+/**
+ * \todo If intra or extra cellular ionic concentration is written
+ * then it requires call to `nrn_wrote_conc`. In C backend this is
+ * implemented in `ion_write_statements()` itself but this is not
+ * handled yet.
+ */
+std::vector<ShadowUseStatement> CodegenInfo::ion_write_statements(BlockType type) {
```
[the remark this answers is not in the document]
Comment (pramodk) [reply src/codegen/codegen_info.cpp:167]: Oh...I forgot to remove them from here! thanks!

**Review by iomaganaris — APPROVED** [review]

**Pull request merged**